src/Sylius/Bundle/CoreBundle/Checkout/Step/SecurityStep.php

```diff
@@ -144,23 +125,77 @@ protected function isGuestOrderAllowed()
      */
     protected function overrideSecurityTargetPath()
     {
-        $providerKey = $this->container->getParameter('fos_user.firewall_name');
+        $this->get('session')->set('_security.main.target_path', $this->generateUrl('sylius_checkout_security', array(), true));
+    }
 
-        $this->get('session')->set('_security.'.$providerKey.'.target_path', $this->generateUrl('sylius_checkout_security', array(), true));
+    /**
+     * @param FormInterface $guestForm
+     * @param OrderInterface $order
+     * @return ActionResult
+     */
+    protected function processGuestOrder(FormInterface $guestForm, OrderInterface $order)
+    {
+        $customer = $guestForm->getData();
+        $order->setCustomer($customer);
+        $this->dispatchCheckoutEvent(SyliusCheckoutEvents::SECURITY_PRE_COMPLETE, $order);
+        $this->saveResource($order);
+
+        return $this->complete();
     }
 
     /**
-     * Dispatch security events, update user and flush
-     *
-     * @param UserInterface $user
+     * @param FormInterface  $registrationForm
+     * @param OrderInterface $order
+     * @return ActionResult
      */
-    protected function saveUser(UserInterface $user)
+    protected function processRegistration(FormInterface $registrationForm, OrderInterface $order)
     {
-        $order = $this->getCurrentCart();
+        $this->registerUser($registrationForm);
         $this->dispatchCheckoutEvent(SyliusCheckoutEvents::SECURITY_PRE_COMPLETE, $order);
+        $this->saveResource($order);
 
-        $this->get('fos_user.user_manager')->updateUser($user, true);
+        return $this->complete();
+    }
+
+    /**
+     * @param OrderInterface $order
+     * @return ActionResult
+     */
+    protected function processUserLoggedIn(OrderInterface $order)
+    {
+        $this->dispatchCheckoutEvent(SyliusCheckoutEvents::SECURITY_PRE_COMPLETE, $order);
+        $this->saveResource($order);
+
+        return $this->complete();
+    }
 
-        $this->dispatchCheckoutEvent(SyliusCheckoutEvents::SECURITY_COMPLETE, $order);
+    /**
+     * @param FormInterface $registrationForm
+     *
+     * @return UserInterface
+     */
+    protected function registerUser(FormInterface $registrationForm)
+    {
+        $customer = $registrationForm->getData();
+        $this->dispatchEvent('sylius.customer.pre_register', new ResourceEvent($customer));
```

Comment:
It could be, but right now this custom register is only performed here, so I'm not sure if creating a separate 3 lines service for that isn't overkill.